### PR TITLE
bad alloc fix for AgentHost

### PR DESCRIPTION
**Pull request by CICS-Oleg**
Fix for bad alloc problem added

### CODE REVIEW

**Review by noskill — CHANGES_REQUESTED** [review]

File: Malmo/src/AgentHost.cpp [hunk 1]
```diff
@@ -128,10 +128,15 @@ namespace malmo
 
     void AgentHost::startMission(const MissionSpec& mission, const MissionRecordSpec& mission_record)
     {
-        ClientPool client_pool;
+        {ClientPool client_pool;
```
[the remark this answers is not in the document]
Comment (CICS-Oleg) [reply Malmo/src/AgentHost.cpp]: There is no { in the last commit.

**Review by noskill — APPROVED** [review]

File: Malmo/src/TimestampedVideoFrame.cpp [hunk 2]
```diff
@@ -58,6 +58,7 @@ namespace malmo
         modelViewMatrix = std::vector<float>(16, 0.0);
 
         // First extract the positional information from the header:
+	//std::cout <<"frame constructor "<< message.data.size()<<std::endl;
```
Comment (noskill) [inline Malmo/src/TimestampedVideoFrame.cpp:61]:
```suggestion
```

File: Malmo/src/TimestampedVideoFrame.cpp [hunk 3]
```diff
@@ -113,6 +114,7 @@ namespace malmo
     std::ostream& operator<<(std::ostream& os, const TimestampedVideoFrame& tsvidframe)
     {
         os << "TimestampedVideoFrame: " << to_simple_string(tsvidframe.timestamp) << ", type " << tsvidframe.frametype << ", " << tsvidframe.width << " x " << tsvidframe.height << " x " << tsvidframe.channels << ", (" << tsvidframe.xPos << "," << tsvidframe.yPos << "," << tsvidframe.zPos << " - yaw:" << tsvidframe.yaw << ", pitch:" << tsvidframe.pitch << ")";
+	//std::cout<<"!!! 1"<<std::endl;
```
Comment (noskill) [inline Malmo/src/TimestampedVideoFrame.cpp:117]:
```suggestion
```

File: Malmo/src/VideoServer.cpp [hunk 4]
```diff
@@ -144,15 +145,18 @@ namespace malmo
...
         this->handle_frame(frame); 
 
         for (const auto& writer : this->writers){
             if (writer->isOpen()){
-                if (writer->write(frame)){
+		//std::cout<<"before writer " << frame->pixels.size()<<std::endl;
+                if (writer->write(*frame)){
                     this->queued_frames++;
                 }
+		//std::cout<<"after writer " << frame->pixels.size()<<std::endl;
```
Comment (noskill) [inline Malmo/src/VideoServer.cpp:159]:
```suggestion
```

File: Malmo/src/TimestampedVideoFrame.cpp [hunk 5]
```diff
@@ -131,6 +132,7 @@ namespace malmo
         default:
             break;
         }
+	//std::cout<<"!!! 2"<<std::endl;
```
Comment (noskill) [inline Malmo/src/TimestampedVideoFrame.cpp:135]:
```suggestion
```

File: Malmo/src/AgentHost.cpp [hunk 6]
```diff
@@ -815,23 +823,25 @@ namespace malmo
...
             }
         }
 
-        if (message.frametype == TimestampedVideoFrame::COLOUR_MAP) {
-            this->world_state.video_frames_colourmap.push_back( boost::make_shared<TimestampedVideoFrame>( message ) );
+        if (message->frametype == TimestampedVideoFrame::COLOUR_MAP) {
+            this->world_state.video_frames_colourmap.push_back( boost::make_shared<TimestampedVideoFrame>( *message ) );
         } else {
-            this->world_state.video_frames.push_back( boost::make_shared<TimestampedVideoFrame>( message ) );
+	//auto v = boost::make_shared<TimestampedVideoFrame>( *message );
```
Comment (noskill) [inline Malmo/src/AgentHost.cpp:842]:
```suggestion
```

File: Malmo/src/AgentHost.cpp [hunk 7]
```diff
@@ -815,23 +823,24 @@ namespace malmo
...
         }
 
-        if (message.frametype == TimestampedVideoFrame::COLOUR_MAP) {
-            this->world_state.video_frames_colourmap.push_back( boost::make_shared<TimestampedVideoFrame>( message ) );
+        if (message->frametype == TimestampedVideoFrame::COLOUR_MAP) {
+            this->world_state.video_frames_colourmap.push_back( boost::make_shared<TimestampedVideoFrame>( *message ) );
         } else {
-            this->world_state.video_frames.push_back( boost::make_shared<TimestampedVideoFrame>( message ) );
+            this->world_state.video_frames.emplace_back( message );
+
```
Comment (noskill) [inline Malmo/src/AgentHost.cpp:843]:
```suggestion
```

File: Malmo/src/AgentHost.cpp [hunk 8]
```diff
@@ -815,23 +823,23 @@ namespace malmo
         this->current_mission_record.reset();
     }
 
-    void AgentHost::onVideo(TimestampedVideoFrame message)
+    void AgentHost::onVideo(std::shared_ptr<TimestampedVideoFrame> message)
     {
         boost::lock_guard<boost::mutex> scope_guard(this->world_state_mutex);
-        LOGSIMPLE(LOG_FINE, "processing video message");
+        LOGSIMPLE(LOG_FINE, "processing video message");
```
Comment (noskill) [inline Malmo/src/AgentHost.cpp:829]:
```suggestion
        LOGSIMPLE(LOG_FINE, "processing video message");
```

File: Malmo/src/AgentHost.cpp [hunk 9]
```diff
@@ -221,6 +228,7 @@ namespace malmo
             std::ofstream missionInitXML(this->current_mission_record->getMissionInitPath());
             missionInitXML << this->current_mission_init->getAsXML(true);
         }
+
```
Comment (noskill) [inline Malmo/src/AgentHost.cpp:231]:
```suggestion
```

File: Malmo/src/AgentHost.cpp [hunk 10]
```diff
@@ -208,10 +213,12 @@ namespace malmo
             // this is a multi-agent mission and we are not the agent that connects to the client with the integrated server
             // so we need to ask the clients in the client pool until we find it
             findServer(pool);
+		
         }
         
         // work through the client pool until we find a client to run our mission for us
         findClient( pool );
+
```
Comment (noskill) [inline Malmo/src/AgentHost.cpp:221]:
```suggestion
```

File: Malmo/src/AgentHost.cpp [hunk 11]
```diff
@@ -208,10 +213,12 @@ namespace malmo
             // this is a multi-agent mission and we are not the agent that connects to the client with the integrated server
             // so we need to ask the clients in the client pool until we find it
             findServer(pool);
+
```
Comment (noskill) [inline Malmo/src/AgentHost.cpp:216]:
```suggestion
```

File: Malmo/src/AgentHost.cpp [hunk 12]
```diff
@@ -159,16 +160,18 @@ namespace malmo
     void AgentHost::startMission(const MissionSpec& mission, const ClientPool& client_pool, const MissionRecordSpec& mission_record, int role, std::string unique_experiment_id)
     {
         std::call_once(test_schemas_flag, testSchemasCompatible);
-
+
```
Comment (noskill) [inline Malmo/src/AgentHost.cpp:163]:
```suggestion
```

File: Malmo/src/AgentHost.cpp [hunk 13]
```diff
@@ -128,7 +128,8 @@ namespace malmo
 
     void AgentHost::startMission(const MissionSpec& mission, const MissionRecordSpec& mission_record)
     {
-        ClientPool client_pool;
+
```
Comment (noskill) [inline Malmo/src/AgentHost.cpp:131]:
```suggestion
```

File: Malmo/src/AgentHost.cpp [hunk 14]
```diff
@@ -159,16 +160,18 @@ namespace malmo
...
         {
             if (mission.getNumberOfAgents() == 1)
                 throw MissionException("Role " + std::to_string(role) + " is invalid for this single-agent mission - must be 0.", MissionException::MISSION_BAD_ROLE_REQUEST);
             else
                 throw MissionException("Role " + std::to_string(role) + " is invalid for this multi-agent mission - must be in range 0-" + std::to_string(mission.getNumberOfAgents() - 1) + ".", MissionException::MISSION_BAD_ROLE_REQUEST);
         }
+		
         if( mission.isVideoRequested( role ) ) 
         {
+
```
Comment (noskill) [inline Malmo/src/AgentHost.cpp:174]:
```suggestion
```

File: Malmo/src/VideoServer.cpp [hunk 15]
```diff
@@ -144,13 +145,15 @@ namespace malmo
             // one when the same port has been reassigned. Could throw here but chose to silently ignore since very rare.
             return;
         }
-        TimestampedVideoFrame frame(this->width, this->height, this->channels, message, this->transform, this->frametype);
+
```
Comment (noskill) [inline Malmo/src/VideoServer.cpp:148]:
```suggestion
```

**Pull request merged**